fs-storage/src/file_storage.rs

```diff
@@ -1,92 +1,101 @@
+use serde::{Deserialize, Serialize};
 use std::fs::{self, File};
-use std::io::{BufRead, BufReader, BufWriter, Write};
+use std::io::{BufWriter, Write};
 use std::time::SystemTime;
 use std::{
     collections::BTreeMap,
     path::{Path, PathBuf},
 };
 
 use crate::base_storage::BaseStorage;
+use crate::utils::read_version_2_fs;
 use data_error::{ArklibError, Result};
 
-const STORAGE_VERSION: i32 = 2;
-const STORAGE_VERSION_PREFIX: &str = "version ";
+/*
+Note on `FileStorage` Versioning:
 
-pub struct FileStorage<K, V> {
+`FileStorage` is a basic key-value storage system that persists data to disk.
+
+In version 2, `FileStorage` stored data in a plaintext format.
+Starting from version 3, data is stored in JSON format.
+
+For backward compatibility, we provide a helper function `read_version_2_fs` to read version 2 format.
+*/
+const STORAGE_VERSION: i32 = 3;
+
+pub struct FileStorage<K, V>
+where
+    K: Ord,
+{
     label: String,
     path: PathBuf,
     timestamp: SystemTime,
+    data: FileStorageData<K, V>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct FileStorageData<K, V>
+where
+    K: Ord,
+{
+    version: i32,
     data: BTreeMap<K, V>,
 }
 
 impl<K, V> FileStorage<K, V>
 where
-    K: Ord + Clone + serde::Serialize + serde::de::DeserializeOwned,
-    V: Clone + serde::Serialize + serde::de::DeserializeOwned,
+    K: Ord
+        + Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::str::FromStr,
+    V: Clone
+        + serde::Serialize
+        + serde::de::DeserializeOwned
+        + std::str::FromStr,
 {
     /// Create a new file storage with a diagnostic label and file path
     pub fn new(label: String, path: &Path) -> Self {
         let mut file_storage = Self {
             label,
             path: PathBuf::from(path),
             timestamp: SystemTime::now(),
-            data: BTreeMap::new(),
+            data: FileStorageData {
+                version: STORAGE_VERSION,
+                data: BTreeMap::new(),
+            },
         };
 
         // Load the data from the file
-        file_storage.data = match file_storage.read_fs() {
+        file_storage.data.data = match file_storage.read_fs() {
```

Comment:
What if we rename the inner `data` field into `entries` or `values`?

Reply:
Renamed to `entries`